File: 02/part1.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin
});

let codes;
rl.on('line', input => {
  codes = input.split(',').map(x => parseInt(x, 10));
  console.log(`\nRead ${codes.length} codes`);
});

const ADD = 1;
const MULT = 2;
const HALT = 99;
function run() {
  console.log('Running...');
  codes[1] = 12;
  codes[2] = 2;
  let ptr = 0;
  while (codes[ptr] !== HALT && ptr < 3000) {
    const op = codes[ptr];
    const arg1 = codes[codes[ptr+1]] || 0;
    const arg2 = codes[codes[ptr+2]] || 0;
    const outptr = codes[ptr+3];
    const out = op === ADD ? arg1 + arg2 : arg1 * arg2;
    // console.log(ptr, op, arg1, arg2, outptr, out);
    codes[outptr] = out;
    ptr += 4;
  }
  console.log(`\nOut: ${codes[0]}`);
}

rl.on('close', run);

File: 05/part1.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func toCodes(vals []string) []int {
	codes := make([]int, len(vals))
	for i, s := range vals {
		codes[i], _ = strconv.Atoi(s)
	}
	return codes
}

func getOpCode(code int) (int, [3]int) {
	var modes [3]int
	op := code % 100
	modes[0] = code / 100 % 10
	modes[1] = code / 1000 % 10
	modes[2] = code / 10000 % 10
	return op, modes
}

const ADD int = 1
const MULT int = 2
const INPUT int = 3
const OUTPUT int = 4
const HALT int = 99

func computer(codes []int, input []int) {
	var (
		inputPtr int
		ptr      int
		opcode   int
		modes    [3]int
	)

	for ptr < len(codes) {
		opcode, modes = getOpCode(codes[ptr])
		switch opcode {
		case ADD:
			arg1 := codes[ptr+1]
			arg2 := codes[ptr+2]
			if modes[0] == 0 {
				arg1 = codes[arg1]
			}
			if modes[1] == 0 {
				arg2 = codes[arg2]
			}
			out := codes[ptr+3]
			codes[out] = arg1 + arg2
			ptr += 4
		case MULT:
			arg1 := codes[ptr+1]
			arg2 := codes[ptr+2]
			if modes[0] == 0 {
				arg1 = codes[arg1]
			}
			if modes[1] == 0 {
				arg2 = codes[arg2]
			}
			out := codes[ptr+3]
			codes[out] = arg1 * arg2
			ptr += 4
		case INPUT:
			arg := codes[ptr+1]
			codes[arg] = input[inputPtr]
			inputPtr++
			ptr += 2
		case OUTPUT:
			arg := codes[ptr+1]
			if modes[0] == 0 {
				arg = codes[arg]
			}
			fmt.Println(arg)
			ptr += 2
		case HALT:
			fmt.Println("halt")
			return
		}
	}
}

func main() {
	var (
		input string
		codes []int
	)
	fmt.Scanln(&input)
	codes = toCodes(strings.Split(input, ","))

	computer(codes, []int{1})
}
